packages/adblocker/src/preprocessor.ts

```diff
@@ -0,0 +1,241 @@
+      const parenthesesOpenIndex = stack.lastIndexOf('(');
+
+      // If the opening parenthesis doesn't exist
+      if (parenthesesOpenIndex < 0) {
+        return false;
+      }
+
+      output.push(...stack.splice(parenthesesOpenIndex + 1, stack.length - parenthesesOpenIndex));
+      stack.pop();
+    } else if (isOperator(token)) {
+      while (
+        stack.length &&
+        isOperator(stack[stack.length - 1] as string) &&
+        precedence[token] <= precedence[stack[stack.length - 1] as string]
+      ) {
+        output.push(stack.pop()!);
+      }
+
+      stack.push(token);
+    } else if (token === 'true') {
+      output.push(true);
+    } else if (token === 'false') {
+      output.push(false);
+    } else {
+      output.push(!!env.get(token));
+    }
+  }
+
+  // Exit if an unallowed character found.
+  // Since we're tokenizing via String.prototype.match function,
+  // the total length of matched tokens will be different in case
+  // unallowed characters were injected.
+  // However, we expect all spaces were already removed in prior step.
+  if (length !== size) {
+    return false;
+  }
+
+  while (stack.length) {
+    output.push(stack.pop()!);
+  }
+
+  for (const token of output) {
+    if (token === true || token === false) {
+      stack.push(token);
+    } else if (isOperator(token)) {
+      if (token === '!') {
+        stack.push(!stack.pop());
+
+        continue;
+      }
+
+      const right = stack.pop()!;
+      const left = stack.pop()!;
+
+      if (token[0] === '&') {
+        stack.push(left && right);
+      } else {
+        stack.push(left || right);
+      }
+    }
+  }
+
+  // Return false if we didn't see any identifiers.
+  if (!stack[0]) {
+    return false;
+  }
+
+  return stack[0] as boolean;
+};
+
+export default class Preprocessor {
+  public static getCondition(line: string) {
+    return line.slice(5 /* '!#if '.length */).replace(/ /g, '');
+  }
+
+  public static parse(condition: string, filters?: Set<number>): Preprocessor {
+    return new this({
+      condition,
+      filters,
+    });
+  }
+
+  public static deserialize(view: StaticDataView): Preprocessor {
+    const condition = view.getUTF8();
+
+    const filters = new Set<number>();
+    for (let i = 0, l = view.getUint32(); i < l; i++) {
+      filters.add(view.getUint32());
+    }
+
+    return new this({
+      condition,
+      filters,
+    });
+  }
+
+  public readonly condition: string;
+  public readonly filters: Set<number>;
+
+  constructor({
+    condition,
+    filters = new Set(),
+  }: {
+    condition: string;
+    filters?: Set<number> | undefined;
+  }) {
+    this.condition = condition;
+    this.filters = filters;
+  }
+
+  public evaluate(env: Env) {
+    return evaluate(this.condition, env);
+  }
+
+  public serialize(view: StaticDataView) {
+    view.pushUTF8(this.condition);
+
+    view.pushUint32(this.filters.size);
+    for (const filter of this.filters) {
+      view.pushUint32(filter);
+    }
```

Comment:
Nit. These are filter IDs.
```suggestion
    for (const filterID of this.filterIDs) {
      view.pushUint32(filterID);
    }
```